### PR TITLE
Spring cleaning: Remove MetaOrObject

### PR DESCRIPTION
**Pull request by ttuegel**
Today I removed many level parameters, and the `MetaOrObject` class itself. I also started to split up the module `Kore.AST.Common`. Most other modules that import this one only need a tiny subset of its definitions. It also takes a long time to compile (up to tens of seconds) because it is so large; GHC can build in parallel if it is split up. Finally, some of the syntax types have their `level` parameter removed and replaced with a `sort` parameter. This is important so we can use a `Sort` type in the syntax checker with more information. It will also be important when we decide to support Applicative Kore, which has no sort information in the parsed source at all. I will finish breaking up `Kore.AST.Common` tomorrow, so it is not essential that this be merged immediately.

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: kore/src/Kore/AST/Sentence.hs [hunk 1]
```diff
@@ -8,7 +8,7 @@ constructs.
 
 Unified constructs are those that represent both meta and object versions of
 an AST term in a single data type (e.g. 'UnifiedSort' that can be either
-'Sort Object' or 'Sort Meta')
+'Sort' or 'Sort')
```
Comment (marick) [inline kore/src/Kore/AST/Sentence.hs]: Now-awkward comment.

File: kore/src/Kore/AST/Common.hs [hunk 2]
```diff
@@ -670,40 +250,40 @@ versions of symbol declarations. It should verify 'MetaOrObject level'.
...
-instance (Ord child, Ord (var lvl)) => Ord (Exists lvl var child) where
+instance (Ord child, Ord variable) => Ord (Exists lvl variable child) where
     compare = compare1
 
-instance (Show child, Show (var lvl)) => Show (Exists lvl var child) where
+instance (Show child, Show variable) => Show (Exists lvl variable child) where
     showsPrec = showsPrec1
 
-instance (Hashable child, Hashable (var lvl)) => Hashable (Exists lvl var child)
+instance (Hashable child, Hashable variable) => Hashable (Exists lvl variable child)
```
Comment (virgil-serbanuta) [inline kore/src/Kore/AST/Common.hs]: Line length

File: kore/src/Kore/AST/Common.hs [hunk 3]
```diff
@@ -783,40 +310,40 @@ versions of symbol declarations. It should verify 'MetaOrObject level'.
...
-instance (Ord child, Ord (var lvl)) => Ord (Forall lvl var child) where
+instance (Ord child, Ord variable) => Ord (Forall lvl variable child) where
     compare = compare1
 
-instance (Show child, Show (var lvl)) => Show (Forall lvl var child) where
+instance (Show child, Show variable) => Show (Forall lvl variable child) where
     showsPrec = showsPrec1
 
-instance (Hashable child, Hashable (var lvl)) => Hashable (Forall lvl var child)
+instance (Hashable child, Hashable variable) => Hashable (Forall lvl variable child)
```
Comment (virgil-serbanuta) [inline kore/src/Kore/AST/Common.hs]: Line length

File: kore/src/Kore/ASTPrettyPrint.hs [hunk 4]
```diff
@@ -211,17 +205,17 @@ inCurlyBracesIndent = listWithDelimiters "{" "}"
 inSquareBracketsIndent :: [Doc ann] -> Doc ann
 inSquareBracketsIndent = listWithDelimiters "[" "]"
 
-instance MetaOrObject level => PrettyPrint (SortVariable level) where
+instance PrettyPrint SortVariable where
     prettyPrint flags sv =
         writeOneFieldStruct flags "SortVariable" (getSortVariable sv)
 
-instance MetaOrObject level => PrettyPrint (Sort level) where
+instance  PrettyPrint Sort where
```
Comment (virgil-serbanuta) [inline kore/src/Kore/ASTPrettyPrint.hs]: Should use a one-space separator (same below).

File: kore/src/Kore/ASTPrettyPrint.hs [hunk 5]
```diff
@@ -303,9 +293,8 @@ instance
             ]
 
 instance
-    ( PrettyPrint child
-    , MetaOrObject level
-    ) => PrettyPrint (Application level child)
+    PrettyPrint child =>
```
Comment (virgil-serbanuta) [inline kore/src/Kore/ASTPrettyPrint.hs]: We should format these consistently (see the instance at line 279, I personally prefer that formatting, which seems more consistent with the other multi-line formatting cases in this file).

File: kore/src/Kore/Step/Rule.hs [hunk 6]
```diff
@@ -113,20 +113,20 @@ rulePattern left right =
...
 
-deriving instance Eq (variable Object) => Eq (RewriteRule level variable)
-deriving instance Ord (variable Object) => Ord (RewriteRule level variable)
-deriving instance Show (variable Object) => Show (RewriteRule level variable)
+deriving instance Eq variable => Eq (RewriteRule level variable)
+deriving instance Ord variable => Ord (RewriteRule level variable)
+deriving instance Show variable => Show (RewriteRule level variable)
 
-instance (Unparse (variable Object), Ord (variable Object)) => Unparse (RewriteRule level variable) where
+instance (Unparse variable, Ord variable) => Unparse (RewriteRule level variable) where
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/Rule.hs]: Line too long (was worse before, though)

**Pull request merged**